### ISSUE TITLE
Questions about results from DeepFillV2 implementation

### ISSUE THREAD

**Issue by ThJOD** [JiahuiYu/generative_inpainting#267]
### Question to DeepFillV2 implementation
Hey,
so I implemented DeepFill2 as well as I could based on your papers and the questions you answered here. 
I trained on a facade dataset for around 6 days on a 1080. 
In addition to groundtruth, image with mask, result and flow I also output the coarse image and the whole image after contextual attention (so x1 and x2).

Here is an example I am quite happy with:
![DeepFillV2_exampleGood](https://user-images.githubusercontent.com/20227707/58471113-b9362180-8175-11e9-8fd3-7174eec799cc.png)
<br>
<br>
But in other cases there are these weird artifacts that the inpainted pixels look like they are not really taken from the picture, but pretty independent, e.g. in the top of this image:
![DeepFillV2_exampleBad](https://user-images.githubusercontent.com/20227707/58471112-b9362180-8175-11e9-9edb-12a35a3ecfd2.png)

So regarding my results I have the following questions:

**Q1**: In the output of the coarse image the mask is still clearly visible and seems to not really have a lot of detail in it, but it still seems just gray. Did your network also learn to just leave the mask area gray in the coarse image?

**Q2**: Do you have an idea what might cause the artifacts in the second image?

### Questions to Extension of DeepfillV2
Additionally to reimplementing DeepfillV2 I want to extend it by not only providing an edge image like you did for the guided inpainting part, but I want to provide a complete segmentation map.

What I do is that I take a segmentation map with 8 classes, then one_hot encode it and attach it to the rgb image as additional channels as an input. I provide segmentation information for the masked out area as well. 
So I attach the segmentation information to the stage1 input, the stage2 input and the discriminator input. 

The problem is that the results during training are quite bad for this use case.
![DeepFillV2_segmentation_exampleBad](https://user-images.githubusercontent.com/20227707/58471905-93118100-8177-11e9-9c5b-6b4cc1cb13b3.png)
This result is after 2.5 days of training on the cityscapes dataset.
<br>
<br>
![DeepFillV2_segmentation_exampleBad2](https://user-images.githubusercontent.com/20227707/58471906-93118100-8177-11e9-99ef-b0c1ae62d231.png)
This result is after 1.5 days on a facade dataset.

As you can see it looks like the contextual attention stage is matching features which are not in the image itself.

My assumption is that since I give the result from coarse as well as the segmentation as input the contextual attention module is "confused" and matches features from the segmentation map. The rgb image has only 3 channels, while the segmentation has 8 channels, therefore it might not focus on the rgb image features.

**Q3**: Do you think in this case it might make sense to not give the segmentation as input to the contextual attention state so that it focuses on image features, or are there any other changes that you think might help?

Thank you very much for reading this long issue with lots of questions.
I am looking forward to your thoughts.

Best regards
ThJOD

**Comment by JiahuiYu** [comment]
Hi,

Thanks for your interest in our work.

First for Q1&Q2, I guess the model has just collapsed. What batch size you are using? Usually large batch size leads to better stability. Moreover, adding tanh/sigmoid activation at the end of generation network may also help on the stability. For all other three examples, it is not likely the contextual attention is matching some background, but more like the network has collapsed.

For Q3, it is a good question. I would suggest not to give segmentation as input to contextual attention. The values in segmentation may not be consistent with values in the learned feature. But I would also encourage you to have a try and do ablation study.

You can reply in this thread if you have more questions.

**Comment by ThJOD** [comment]
Hey,
<br>
thank you for your input, it helped me a lot so far. Just to fill you in on my findings.
Regarding Q1&Q2: I was using batch size 8, but I started using 16 recently, depending on which GPU I have access to, but I did not see any significant difference although I never trained the exact same model with both batchsizes.

I think I figured out what caused the outputs from the x1stage to look so gray and not fill in the masked out area. I changed my implementation of the SN-PatchGAN discriminator to use the SN-Convolution I found in the Neuralgym-Dev branch and instead of using 1:1 for L1 and g_loss I used a factor of 0.1 for g_loss. One result image can be seen here: 
![DeepFillV2_exampleGood2 png](https://user-images.githubusercontent.com/20227707/59079888-6f3d0080-8918-11e9-97d3-2b7f685cc28d.png)
I think the factor for the g_loss played the biggest role in the improvement, but I am not a 100% sure about that.
<br>

Regarding Q3: With the changes mentioned above it worked way better with segmentation as an input as well. I first tested it without giving segmentation to x2 as input and the artifacts only appeared if the mask had a certain shape:
![DeepFillV2_segmentation_exampleGood1](https://user-images.githubusercontent.com/20227707/59079959-ccd14d00-8918-11e9-8f38-4c1558566588.png)
![DeepFillV2_segmentation_exampleBad3](https://user-images.githubusercontent.com/20227707/59080146-e6bf5f80-8919-11e9-8e93-5970bc780f13.png)

Only if the mask had this 'Tree' shape the artifact appeared, but not even then all the time.

**Q4**: Do you think there might be a certain Mask shape which is especially problematic for the contextual attention layer? E.g. Really long strokes (More than half the image high), or ellipse shaped masks?

I then trained a model which took segmentation information as input for the x2 stage. It seems that the artifacts appear less often then. But since I use different kinds of masks each batch randomly and their shape is random as well it might just be that the training is highly influenced by what random mask is used.

Best regards
ThJOD

**Comment by JiahuiYu** [comment]
@ThJOD Thanks for your feedback.

For Q4, contextual attention could work on long strokes or ellipse masks. Our implementation takes mask as input, so as long as you are training with certain kinds of masks, the inference performance should not be bad.

**Comment by ThJOD** [comment]
Just another update:
Due to my inexperience with tensorflow the masks that I generated we actually not random and the network did not see the ellipse shaped masks during training. That is why these artifacts appeared.
Best regards and thanks for your tips
ThJOD

**Comment by ThJOD** [comment]
Hey @JiahuiYu ,
I have another question regarding some results from my DeepfillV2 implementation.
I noticed that the inpainted images appear blurry on the lower and right edge of the image.
For example:
When I let my network inpaint this image with the following mask the result is this:
![facade_010](https://user-images.githubusercontent.com/20227707/63148488-ee933280-c033-11e9-9bbf-2f33c4851a83.png)
![facade_010_mask](https://user-images.githubusercontent.com/20227707/63148492-f18e2300-c033-11e9-89a9-a136f7eebfdc.png)
![facade_010_orig_single](https://user-images.githubusercontent.com/20227707/63148508-feab1200-c033-11e9-9aef-341db5ab44f8.png)
You can see that the right side is blurry, while the left side is not.
I then flipped the inputs vertically as well as horizontally and did the inpainting again. After inpainting I flipped the image into the original position The result is the following image: 
![facade_010_orig_single](https://user-images.githubusercontent.com/20227707/63148604-4762cb00-c034-11e9-94e6-844cc51f6bd0.png)
You can see that this time the blurriness exists on the opposite side.

I tested the same thing with DeepfillV1 with the provided places2 weights. I could not detect any blurriness on either side, but the images look different depending on whether they were flipped, which is understandable.

Through some other configurations I was able to get rid of the blurriness and now suspect that by adapting the loss function e.g. adding other loss values or maybe reducing l1 loss you can get rid of the blurriness.

But still I am wondering why the blurriness only appears on the right and lower edge of the images.
One assumption is that it has to do with the uneven padding of tensorflow with 'SAME' (If even padding is not possible it will pad more on the right and bottom). And this way the tensors have more 0s there and the result is more blurry.

**Q5**
Have you also experienced uneven blurriness and what are your thoughts on the reason behind it?

Best regards
ThJOD

**Comment by JiahuiYu** [comment]
Hi, that is indeed a good finding. And thanks for your clear explanation. I never notice this effect before because usually the mask I tested are not at the boundary.

I think the problem is not because of the padding. To test this assumption, if you carefully choose the image size (divisible by 32), the padding can be even. If in that case the performance is good. Then you are right.

More likely, in my guess, would be just the randomness of the trained model. This checkpoint happens to be very sensitive to right masks.

**Comment by ThJOD** [comment]
The result is the same if the images and masks have been resized to 256x256. You are right then the padding should be the same on both sides for all convolutions. 

Thank you for the input!

**Comment by fzhang612** [comment]
@ThJOD Just curious based on what reason/observation you changed L1 to g_loss ratio from 1 : 1 to 1 : 0.1. Thanks!

**Comment by ThJOD** [comment]
> @ThJOD Just curious based on what reason/observation you changed L1 to g_loss ratio from 1 : 1 to 1 : 0.1. Thanks!

I do not remember exactly what gave me the idea to try that out anymore. I think looking in tensorboard at the loss values from DeepFillV1 and the loss values that I received with my DeepFillV2 adaption I noticed that the ratio is not the same, so I tried out something new. 

And I had to try something, because the inpainting did not work. Also you could see that the images after stage 1 were still quite grayish and intuitively if the l1 loss is more dominant it should reduce that and introduce more color. 
I think I got to 0.1 just by trial and error. It might be, that it is not the optimal ratio. 

Another example where someone used a DeepFillV2 adaption, but adapted the ratio can be seen here: **https://arxiv.org/abs/1902.06838** but I only found this afterwards.

**Comment by fzhang612** [comment]
Thanks @ThJOD, for your response.

In my case, I notice, from tensorboard outputs that partial derivative of g_loss wrt x1 and x2 is roughly 4 times bigger than that of l1_loss wrt x1 and x2. In order to make the 2 derivatives comparable, I weight g_loss by 1/4. Does this sound a reasonable approach?

**Comment by ThJOD** [comment]
From my limited knowledge about Neural Networks this sounds like a good first step to make sure that you have a good starting point where no loss overpowers the other. But as many papers have stated each loss affects the network differently, e.g. the l1 loss will make the result more blurry, but on average it looks ok, while the adversarial loss will be able to add more detail, but can produce artifacts. So it might make sense to start out with 1/4 and then if you have enough time try out other ratios to see if it improves the results for your use case.

**Comment by JiahuiYu** [comment]
@ThJOD Thanks for your help to others!
